core/chain-configs/src/test_genesis.rs

```diff
@@ -68,229 +92,236 @@ struct UserAccount {
+    pub fn target_validator_mandates_per_shard(
+        mut self,
+        target_validator_mandates_per_shard: NumSeats,
+    ) -> Self {
+        self.target_validator_mandates_per_shard = Some(target_validator_mandates_per_shard);
         self
     }
 
-    pub fn epoch_length(&mut self, epoch_length: BlockHeightDelta) -> &mut Self {
-        self.epoch_length = Some(epoch_length);
+    pub fn shuffle_shard_assignment_for_chunk_producers(mut self, shuffle: bool) -> Self {
+        self.shuffle_shard_assignment_for_chunk_producers = Some(shuffle);
         self
     }
 
-    pub fn shard_layout_single(&mut self) -> &mut Self {
-        self.epoch_config_mut().shard_layout = ShardLayout::single_shard();
+    // Validators with performance below 80% are kicked out, similarly to
+    // mainnet as of 28 Jun 2024.
+    pub fn kickouts_standard_80_percent(mut self) -> Self {
+        self.block_producer_kickout_threshold = Some(80);
+        self.chunk_producer_kickout_threshold = Some(80);
+        self.chunk_validator_only_kickout_threshold = Some(80);
         self
     }
 
-    pub fn shard_layout_simple_v1(&mut self, boundary_accounts: &[&str]) -> &mut Self {
-        let boundary_accounts = boundary_accounts.iter().map(|a| a.parse().unwrap()).collect();
-        self.epoch_config_mut().shard_layout =
-            ShardLayout::multi_shard_custom(boundary_accounts, 1);
+    // Only chunk validator-only nodes can be kicked out.
+    pub fn kickouts_for_chunk_validators_only(mut self) -> Self {
+        self.block_producer_kickout_threshold = Some(0);
+        self.chunk_producer_kickout_threshold = Some(0);
+        self.chunk_validator_only_kickout_threshold = Some(50);
         self
     }
 
-    // TODO(#11265): move this and relevant methods to epoch config builder.
-    // In dynamic resharding world, shard layout will not be static.
-    pub fn shard_layout(&mut self, shard_layout: ShardLayout) -> &mut Self {
-        self.epoch_config_mut().shard_layout = shard_layout;
-        self
+    pub fn build(self) -> EpochConfig {
+        let epoch_length = self.epoch_length.unwrap_or(5);
+        let shard_layout = self.shard_layout.unwrap_or_else(ShardLayout::single_shard);
+        let validators_spec =
+            self.validators_spec.unwrap_or_else(|| ValidatorsSpec::DesiredRoles {
+                block_and_chunk_producers: vec!["validator0".to_string()],
+                chunk_validators_only: vec![],
+            });
+        let DerivedValidatorSetup {
+            validators: _,
+            num_block_producer_seats,
+            num_chunk_producer_seats,
+            num_chunk_validator_seats,
+        } = derive_validator_setup(validators_spec);
+
+        let mut epoch_config =
+            Genesis::test_epoch_config(num_block_producer_seats, shard_layout, epoch_length);
+        epoch_config.num_block_producer_seats = num_block_producer_seats;
+        epoch_config.block_producer_kickout_threshold = 0;
```

Comment:
Now that I look at this again, it seems like maybe we should bake some of the defaults into the builders rather than doing it this ad hoc way.

What is this warning ceremony useful for (given this is test-only)? I mean I sort of understand, but we might as well just log the entire configurations instead of one warning for each default being used

Reply:
> should bake some of the defaults into the builders rather than doing it this ad hoc way

I'll do it in a follow-up :))